### PR TITLE
Add mount options to volume plugin API

### PR DESCRIPTION
**Pull request by paulcwarren**
**\- What I did**
We added volume mount options to the volume plugin API and surfaced that to the cli user by extending `docker run`'s volume mount spec.

**\- How to verify it**
Inside the development container with these pull request changes:-
- Build the binaries and run the docker daemon as per the development guidelines
- Go get our "test" localdriver (that we added a mount option to) 

```
go get github.com/cloudfoundry-incubator/localdriver/...
```
- Run the localdriver

```
 mkdir -p /run/docker/plugins/ && localdriver -listenAddr=/run/docker/plugins/localdriver.sock -transport=unix &
```
- Use docker to create a volume (the options here is setting up a lookie likie basic auth, the point being we will have to pass the passcode as a mount option

```
docker volume create --driver=localdriver --name local-volume --opt volume_id=local-volume --opt passcode=topsecret
```
- Run a docker container that touches a file on the created volume (but don't set the passcode) in order to see it fail

```
docker run --name hello-world -v local-volume:/data:rw --volume-driver=localdriver busybox /bin/sh -c "echo Hellopersistentworld > /data/hello.world"
```

You should see it fail with output like this:-

```
DEBU[56925] Assigning addresses for endpoint hello-world2's interface on network bridge 
DEBU[56925] Programming external connectivity on endpoint hello-world2 (83b536c04badb6e00c48c662a0f8e9517241e93a7cbff27a2803cd7b5b9b4509) 
{"timestamp":"1469201235.981245756","source":"local-driver-server","message":"local-driver-server.server.handle-mount.start","log_level":1,"data":{"session":"1.9"}}
{"timestamp":"1469201235.981326580","source":"local-driver-server","message":"local-driver-server.server.handle-mount.mount.missing-passcode","log_level":1,"data":{"session":"1.9.1","volume":"local-volume","volume_name":"local-volume"}}
{"timestamp":"1469201235.981382847","source":"local-driver-server","message":"local-driver-server.server.handle-mount.failed-mounting-volume","log_level":2,"data":{"error":"Volume local-volume requires a passcode","session":"1.9","volume":"local-volume"}}
{"timestamp":"1469201235.981415272","source":"local-driver-server","message":"local-driver-server.server.handle-mount.end","log_level":1,"data":{"session":"1.9"}}
DEBU[56925] Revoking external connectivity on endpoint hello-world2 (83b536c04badb6e00c48c662a0f8e9517241e93a7cbff27a2803cd7b5b9b4509) 
DEBU[56925] Releasing addresses for endpoint hello-world2's interface on network bridge 
DEBU[56925] ReleaseAddress(LocalDefault/172.18.0.0/16, 172.18.0.2) 
{"timestamp":"1469201236.102541924","source":"local-driver-server","message":"local-driver-server.server.handle-unmount.start","log_level":1,"data":{"session":"1.10"}}
{"timestamp":"1469201236.102647305","source":"local-driver-server","message":"local-driver-server.server.handle-unmount.unmount.getting-volume","log_level":1,"data":{"name":"local-volume","session":"1.10.1","volume":"local-volume"}}
{"timestamp":"1469201236.102695704","source":"local-driver-server","message":"local-driver-server.server.handle-unmount.unmount.failed-mountpoint-not-assigned","log_level":2,"data":{"error":"Volume not previously mounted","session":"1.10.1","volume":"local-volume"}}
{"timestamp":"1469201236.102714777","source":"local-driver-server","message":"local-driver-server.server.handle-unmount.failed-unmount-volume","log_level":2,"data":{"error":"Volume not previously mounted","session":"1.10","volume":"local-volume"}}
{"timestamp":"1469201236.102750301","source":"local-driver-server","message":"local-driver-server.server.handle-unmount.end","log_level":1,"data":{"session":"1.10"}}
WARN[56925] 2bb53fc58e961f13582217ee6fe2844a494b1c010ef8af65466db3f7e3390c3d cleanup: Failed to umount volumes: Volume not previously mounted 
ERRO[56925] Handler for POST /v1.25/containers/2bb53fc58e961f13582217ee6fe2844a494b1c010ef8af65466db3f7e3390c3d/start returned error: Volume local-volume requires a passcode 
DEBU[56925] attach: stderr: end                          
DEBU[56925] attach: stdout: end
```
- Cat hello.world to ensure it doesn't exist

```
cat /tmp/volumes/_volumes/local-volume/hello.world
cat: /tmp/volumes/_volumes/local-volume/hello.world: No such file or directory
```
- Re-run this docker container this time setting the `passcode` mount option

```
docker run --name hello-world2 -v local-volume:/data:rw:passcode=topsecret --volume-driver=localdriver busybox /bin/sh -c "echo Hellopersistentworld > /data/hello.world"
```
- cat hello.world to ensure it wrote to the file this time

```
cat /tmp/volumes/_volumes/local-volume/hello.world
Hellopersistentworld
```

_Notes_
- We want to add a cli_integration test for this new feature.  Were originally thought we would wire mount options in the existing local volume (`volume/local`).  However, on examining that code we realized that a local mount is actuvally performed as part of `docker volume create`, not `docker run`.   We then thought about 'mocking' a volume plugin.  We found [this](https://github.com/docker/go-plugins-helpers/blob/master/volume/shim/shim_test.go) which seemed like it could wrap testutils.FakeDriver (or something similar).  However, this wasn't an existing dependency and didn't comply with the docker vendoring guidelines so we decided not to add it and backed out of this alternative too.  

cute animal

Signed-off-by: Paul Warren paul.warren@emc.com
Signed-off-by: Luke Woydziak luke.woydziak@emc.com
Signed-off-by: Connor Braa cbraa@pivotal.io
Signed-off-by: Julian Hjortshoj julian.hjortshoj@emc.com

### CODE REVIEW

File: volume/volume.go [hunk 1]
```diff
@@ -58,7 +58,7 @@ type Volume interface {
 	Path() string
 	// Mount mounts the volume and returns the absolute path to
 	// where it can be consumed.
```
Comment (AkihiroSuda) [inline volume/volume.go]: I think it is better to add a comment like "opts can be nil"

Comment (cwlbraa) [inline volume/volume.go]: Sure, and we'll throw a similar comment on the Driver.Create interface.

**Comment by cwlbraa** [comment]
Any suggestions as to how one might go about debugging that failing windows build?

**Comment by thaJeztah** [comment]
I'm not really sure about this; first of all it's adding more options to an already overloaded syntax for `--volume`; we really shouldn't add more options to that.

Also, adding access control here doesn't feel right to me; the same could probably be achieved with an Authorization plugin; have you considered using that?

**Comment by cpuguy83** [comment]
I agree on not wanting to add support for arbitrary options on the `-v` syntax, though I certainly see the use case.

**Comment by tedsuo** [comment]
@cpuguy83 we're more interested in the change to the Volume Plugin API, in order to allow options to be passed to `Mount`. (https://github.com/docker/docker/issues/24382). 

Currently, `Create` is the only call which takes options, and this forces driver design to overload the `Create` command. We'd rather have the ability to pass options to `Mount` instead. This is not something that comes up as a problem in the stand-alone/CLI usecase, but does come up when orchestrating a cluster (Create is called once, Mount is called many times on many machines). Does this dovetail with your plans for drivers, now that you're working on orchestration? When we last spoke with @samalba and @mgoelzer, you mentioned that this was a direction you were interested in.

Another way of putting it is that we don't have an opinion about how the CLI should expose this, our focus is on changing the API layer to enable better orchestration:

- `Create` does not need be be called when all the system wants to do is `Mount`.
- Drivers are not forced to supply a backplane/db layer for storing mount directives when the orchestration layer may already have this information.

Happy to take design direction.

Cheers,
Ted

**Comment by thaJeztah** [comment]
We discussed this and we don't want to overload this flag even more; perhaps we should think of adding `--mount` to docker run, but maybe you can do an alternative proposal on design?

I'm closing this for now, but feel free to continue the discussion

**Pull request closed without merging**